File: lib/services/company_service.dart
// Flutter service for company-related API calls
import 'dart:convert';
import 'package:http/http.dart' as http;
import '../models/company.dart';
import '../utils/api_config.dart';

/// Service class for company-related API operations
class CompanyService {
  final http.Client _httpClient;
  final String _baseUrl;
  
  CompanyService({
    http.Client? httpClient,
    String? baseUrl,
  }) : _httpClient = httpClient ?? http.Client(),
       _baseUrl = baseUrl ?? ApiConfig.baseUrl;
  
  /// Get all companies
  Future<List<Company>> getAllCompanies() async {
    final response = await _httpClient.get(
      Uri.parse('$_baseUrl/api/companies'),
      headers: ApiConfig.headers,
    );
    
    if (response.statusCode == 200) {
      final List<dynamic> data = json.decode(response.body);
      return data.map((json) => Company.fromJson(json)).toList();
    } else {
      throw Exception('Failed to load companies: ${response.body}');
    }
  }
  
  /// Get company by ID
  Future<Company> getCompanyById(String id) async {
    final response = await _httpClient.get(
      Uri.parse('$_baseUrl/api/companies/$id'),
      headers: ApiConfig.headers,
    );
    
    if (response.statusCode == 200) {
      return Company.fromJson(json.decode(response.body));
    } else {
      throw Exception('Failed to load company: ${response.body}');
    }
  }
  
  /// Search companies by name
  Future<List<Company>> searchCompanies(String query) async {
    final response = await _httpClient.get(
      Uri.parse('$_baseUrl/api/companies/search?q=$query'),
      headers: ApiConfig.headers,
    );
    
    if (response.statusCode == 200) {
      final List<dynamic> data = json.decode(response.body);
      return data.map((json) => Company.fromJson(json)).toList();
    } else {
      throw Exception('Failed to search companies: ${response.body}');
    }
  }
  
  /// Filter companies by location
  Future<List<Company>> getCompaniesByLocation({
    String? country,
    String? city,
  }) async {
    final queryParams = <String, String>{};
    if (country != null) queryParams['country'] = country;
    if (city != null) queryParams['city'] = city;
    
    final uri = Uri.parse('$_baseUrl/api/companies/location')
        .replace(queryParameters: queryParams);
    
    final response = await _httpClient.get(
      uri,
      headers: ApiConfig.headers,
    );
    
    if (response.statusCode == 200) {
      final List<dynamic> data = json.decode(response.body);
      return data.map((json) => Company.fromJson(json)).toList();
    } else {
      throw Exception('Failed to filter companies by location: ${response.body}');
    }
  }
  
  /// Filter companies by service category
  Future<List<Company>> getCompaniesByCategory(ServiceCategory category) async {
    final response = await _httpClient.get(
      Uri.parse('$_baseUrl/api/companies/category/${category.name.toUpperCase()}'),
      headers: ApiConfig.headers,
    );
    
    if (response.statusCode == 200) {
      final List<dynamic> data = json.decode(response.body);
      return data.map((json) => Company.fromJson(json)).toList();
    } else {
      throw Exception('Failed to filter companies by category: ${response.body}');
    }
  }
  
  /// Get featured companies
  Future<List<Company>> getFeaturedCompanies({int limit = 10}) async {
    final response = await _httpClient.get(
      Uri.parse('$_baseUrl/api/companies/featured?limit=$limit'),
      headers: ApiConfig.headers,
    );
    
    if (response.statusCode == 200) {
      final List<dynamic> data = json.decode(response.body);
      return data.map((json) => Company.fromJson(json)).toList();
    } else {
      throw Exception('Failed to load featured companies: ${response.body}');
    }
  }
  
  /// Get top-rated companies
  Future<List<Company>> getTopRatedCompanies({
    int minReviews = 5,
    int limit = 10,
  }) async {
    final response = await _httpClient.get(
      Uri.parse('$_baseUrl/api/companies/top-rated?minReviews=$minReviews&limit=$limit'),
      headers: ApiConfig.headers,
    );
    
    if (response.statusCode == 200) {
      final List<dynamic> data = json.decode(response.body);
      return data.map((json) => Company.fromJson(json)).toList();
    } else {
      throw Exception('Failed to load top-rated companies: ${response.body}');
    }
  }
}

File: backend/src/main/kotlin/com/tripbook/catalog/service/CompanyService.kt
package com.tripbook.catalog.service

import com.tripbook.catalog.model.Company
import com.tripbook.catalog.model.ServiceCategory
import java.util.UUID

/**
 * Service interface for business logic related to company management.
 * Defines methods for CRUD operations and queries related to companies.
 */
interface CompanyService {
    /**
     * Retrieves all companies.
     * 
     * @return List of all companies
     */
    suspend fun getAllCompanies(): List<Company>
    
    /**
     * Retrieves a single company by ID.
     * 
     * @param id The company ID
     * @return The company if found, null otherwise
     */
    suspend fun getCompanyById(id: UUID): Company?
    
    /**
     * Creates a new company.
     * 
     * @param company The company to create
     * @return The created company
     */
    suspend fun createCompany(company: Company): Company
    
    /**
     * Updates an existing company.
     * 
     * @param id The ID of the company to update
     * @param company The updated company data
     * @return The updated company if found, null otherwise
     */
    suspend fun updateCompany(id: UUID, company: Company): Company?
    
    /**
     * Deletes a company by ID.
     * 
     * @param id The ID of the company to delete
     * @return true if deleted, false if not found
     */
    suspend fun deleteCompany(id: UUID): Boolean
    
    /**
     * Searches for companies by name.
     * 
     * @param query The search query string
     * @return List of matching companies
     */
    suspend fun searchCompanies(query: String): List<Company>
    
    /**
     * Gets companies by location.
     * 
     * @param country Optional country filter
     * @param city Optional city filter
     * @return List of companies matching the location criteria
     */
    suspend fun getCompaniesByLocation(country: String? = null, city: String? = null): List<Company>
    
    /**
     * Gets companies that offer a specific service category.
     * 
     * @param category The service category to filter by
     * @return List of matching companies
     */
    suspend fun getCompaniesByServiceCategory(category: ServiceCategory): List<Company>
    
    /**
     * Gets featured companies.
     * 
     * @param limit Maximum number of companies to return
     * @return List of featured companies
     */
    suspend fun getFeaturedCompanies(limit: Int = 10): List<Company>
    
    /**
     * Gets top-rated companies.
     * 
     * @param minReviews Minimum number of reviews required
     * @param limit Maximum number of companies to return
     * @return List of top-rated companies
     */
    suspend fun getTopRatedCompanies(minReviews: Int = 5, limit: Int = 10): List<Company>
    
    /**
     * Marks a company as verified.
     * 
     * @param id The ID of the company to verify
     * @return The updated company if found, null otherwise
     */
    suspend fun verifyCompany(id: UUID): Company?
    
    /**
     * Marks a company as featured or removes featured status.
     * 
     * @param id The ID of the company
     * @param featured Whether the company should be featured
     * @return The updated company if found, null otherwise
     */
    suspend fun setCompanyFeaturedStatus(id: UUID, featured: Boolean): Company?
}